### PR TITLE
Upgrade to JUnit 4.13.1. Fixes #827

### LINKED ISSUE #827: Upgrade to JUnit 4.13.
JUnit 4.13 adds two variants of assertThrows.

```
public static <T extends Throwable> T assertThrows(Class<T> expectedThrowable, ThrowingRunnable runnable)
public static <T extends Throwable> T assertThrows(String message, Class<T> expectedThrowable, ThrowingRunnable runnable)
```

These methods clash with `Verify.assertThrows()`.
```
public static void assertThrows(Class<? extends Exception> expectedExceptionClass, Runnable code)
```

This causes some calls to `Verify.assertThrows()` to fail with `java: error: reference to assertThrows is ambiguous`

I wanted to discuss our options here.
* Delete `Verify.assertThrows(Class, Runnable)` in the next major version.
* Or change `Verify` to no longer extend `Assert` in the next major version.
* Deprecate `Verify.assertThrows(Class, Runnable)` now, although it's unclear if the deprecation is actionable.
* Replace usages of `Verify.assertThrows` with `Assert.assertThrow` though it's unclear if that can be done in a minor version.
* Upgrade to JUnit 4.13, though it's unclear if that can be done in a minor version.
* Add `Verify.assertThrows(Class, ThrowingRunnable)` using JUnit's `ThrowingRunnable` for better compatibility in a minor release.

Thoughts @donraab @nikhilnanivadekar?

### PR DESCRIPTION
**Pull request by nikhilnanivadekar**
I have broken the PR down in separate commits to make it easy for the review. Most of the changes are straightforward.

Please note, I have purposely not updated the copyright to make it easy for the review. We are almost in November now, and my plan is that in January I will update all copyright headers to the new year to make it easy for the contributors next year.

### CODE REVIEW

File: eclipse-collections-forkjoin/src/test/java/org/eclipse/collections/impl/forkjoin/FJIterateTest.java [hunk 1]
```diff
@@ -265,7 +267,7 @@ public void testForEachImmutable()
     @Test
     public void testForEachWithException()
     {
-        Verify.assertThrows(RuntimeException.class, () -> FJIterate.forEach(
+        assertThrows(RuntimeException.class, () -> FJIterate.forEach(
```
Comment (prathasirisha) [inline eclipse-collections-forkjoin/src/test/java/org/eclipse/collections/impl/forkjoin/FJIterateTest.java:270]: Suggestion to make this `Assert.assertThrows` instead of a static import, so it is consistent with the rest of the assertions in this class. Same comment for lines 331 and 495.

**Review by prathasirisha — COMMENTED** [review]
Noticed a file called version.properties has junit version (4.12) in it, wanted to bring that to your attention. Assuming that too needs to be updated.

File: eclipse-collections-testutils/src/test/java/org/eclipse/collections/impl/test/AssertNotEqualsThrowsTest.java [hunk 2]
```diff
@@ -38,7 +42,7 @@ public void objectObject()
         }
         catch (AssertionError e)
         {
-            Verify.assertContains(AssertNotEqualsThrowsTest.class.getName(), e.getStackTrace()[0].toString());
+            Verify.assertContains("org.junit.Assert.fail", e.getStackTrace()[0].toString());
```
Comment (donraab) [inline eclipse-collections-testutils/src/test/java/org/eclipse/collections/impl/test/AssertNotEqualsThrowsTest.java:45]: Make the string Static final.

File: eclipse-collections-testutils/src/test/java/org/eclipse/collections/impl/test/VerifyTest.java [hunk 3]
```diff
@@ -475,7 +475,7 @@ public void assertNotEqualsFailsOnSameReference()
         }
         catch (AssertionError e)
         {
-            Verify.assertContains(VerifyTest.class.getName(), e.getStackTrace()[0].toString());
+            Verify.assertContains("org.junit.Assert.fail", e.getStackTrace()[0].toString());
```
Comment (donraab) [inline eclipse-collections-testutils/src/test/java/org/eclipse/collections/impl/test/VerifyTest.java:478]: Make the String static final.

File: unit-tests/unit-tests.iml [hunk 4]
```diff
@@ -11,12 +11,15 @@
...
+    <orderEntry type="library" scope="TEST" name="Maven: org.slf4j:slf4j-nop:1.7.25" level="project" />
     <orderEntry type="module" module-name="eclipse-collections-api" scope="TEST" />
     <orderEntry type="module" module-name="eclipse-collections" scope="TEST" />
     <orderEntry type="module" module-name="eclipse-collections-testutils" scope="TEST" />
-    <orderEntry type="library" scope="TEST" name="Maven: junit:junit:4.12" level="project" />
+    <orderEntry type="library" scope="TEST" name="Maven: junit:junit:4.13.1" level="project" />
     <orderEntry type="library" scope="TEST" name="Maven: org.hamcrest:hamcrest-core:1.3" level="project" />
-    <orderEntry type="library" scope="TEST" name="Maven: org.slf4j:slf4j-api:1.7.25" level="project" />
-    <orderEntry type="library" scope="TEST" name="Maven: org.slf4j:slf4j-nop:1.7.25" level="project" />
+    <orderEntry type="library" scope="TEST" name="Maven: org.slf4j:slf4j-api:1.7.30" level="project" />
```
Comment (donraab) [inline unit-tests/unit-tests.iml:22]: Why is this changed, but the ones above for slf4j are not?

**Review by donraab — CHANGES_REQUESTED** [review]
This PR looks almost ready to go. I have a couple comments inline.

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/IterableTestCase.java [hunk 5]
```diff
@@ -141,8 +141,7 @@ static void assertEquals(Object o1, Object o2)
                     assertSame(comparator1.getClass(), comparator2.getClass());
                 }
             }
-        }
-        else if (o1 instanceof SortedMap<?, ?> || o2 instanceof SortedMap<?, ?>)
+        } else if (o1 instanceof SortedMap<?, ?> || o2 instanceof SortedMap<?, ?>)
```
Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/IterableTestCase.java]: Whitespace nit here

**Review by motlin — APPROVED** [review]

File: eclipse-collections-forkjoin/src/test/java/org/eclipse/collections/impl/forkjoin/FJIterateTest.java (hunk 1 above)
Comment (nikhilnanivadekar) [reply eclipse-collections-forkjoin/src/test/java/org/eclipse/collections/impl/forkjoin/FJIterateTest.java:270]: We have other such instances where we us static import. I dont understand the consistency aspect here. That said, I have fixed it.

File: eclipse-collections-testutils/src/test/java/org/eclipse/collections/impl/test/AssertNotEqualsThrowsTest.java (hunk 2 above)
Comment (nikhilnanivadekar) [reply eclipse-collections-testutils/src/test/java/org/eclipse/collections/impl/test/AssertNotEqualsThrowsTest.java:45]: How does that matter in this test? This test is deprecated as well.

File: eclipse-collections-testutils/src/test/java/org/eclipse/collections/impl/test/VerifyTest.java (hunk 3 above)
Comment (nikhilnanivadekar) [reply eclipse-collections-testutils/src/test/java/org/eclipse/collections/impl/test/VerifyTest.java:478]: How does that matter in a test?

File: unit-tests/unit-tests.iml (hunk 4 above)
Comment (nikhilnanivadekar) [reply unit-tests/unit-tests.iml:22]: will revert this

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/IterableTestCase.java (hunk 5 above)
Comment (nikhilnanivadekar) [reply unit-tests-java8/src/test/java/org/eclipse/collections/test/IterableTestCase.java]: This was fixed in a subsequent commit.

File: unit-tests/unit-tests.iml (hunk 4 above)
Comment (nikhilnanivadekar) [reply unit-tests/unit-tests.iml]: Reverted

**Comment by nikhilnanivadekar** [comment]
@prathasirisha I have fixed the static import and removed the version.properties file as a separate commit. It was needed for ant builds which we dont do anymore.
@donraab I have reverted the iml file. I quite disagree with the static string comment, it is cumbersome in a test to infer such strings. I would prefer not to do that.

**Review by donraab — APPROVED** [review]
LGTM.
